src/Plugin/Field/FieldWidget/OgComplex.php

```diff
@@ -26,13 +31,314 @@
+    $user_group_ids = array_map(function($group) {
+      return $group->id();
+    }, $user_groups_target_type);
+
+    $other_groups_weight_delta = round(count($user_groups) / 2);
+
+    foreach ($items->referencedEntities() as $group) {
+      if (in_array($group->id(), $user_group_ids)) {
+        continue;
+      }
+
+      $elements[$delta] = $this->otherGroupsSingle($delta, $group, $other_groups_weight_delta);
+      $delta++;
+    }
+
+    if (!$form_state->get('other_group_delta')) {
+      $form_state->set('other_group_delta', $delta);
+    }
+
+    // Get the trigger element and check if this the add another item button.
+    $trigger_element = $form_state->getTriggeringElement();
+
+    if ($trigger_element['#name'] == 'add_another_group') {
+      // Increase the number of other groups.
+      $delta = $form_state->get('other_group_delta') + 1;
+      $form_state->set('other_group_delta', $delta);
+    }
+
+    // Add another auto complete field.
+    for ($i = $delta; $i <= $form_state->get('other_group_delta'); $i++) {
+      // Also add one to the weight delta, just to make sure.
+      $elements[$i] = $this->otherGroupsSingle($i, NULL, $other_groups_weight_delta + 1);
+    }
+
+    return $elements;
+  }
+
+  /**
+   * Generating other groups auto complete element.
+   *
+   * @param $delta
+   *   The delta of the new element. Need to be the last delta in order to be
+   *   added in the end of the list.
+   * @param EntityInterface|NULL $entity
+   *   The entity object.
+   * @return array
+   *   A single entity reference input.
+   */
+  public function otherGroupsSingle($delta, EntityInterface $entity = NULL, $weight_delta = 10) {
+    return [
+      'target_id' => [
+        // @todo Allow this to be configurable with a widget setting.
+        '#type' => 'entity_autocomplete',
+        '#target_type' => $this->fieldDefinition->getTargetEntityTypeId(),
+        // todo: fix the definition in th UI level.
+        '#selection_handler' => 'default:og',
+        '#selection_settings' => [
+          'other_groups' => TRUE,
+          'field_mode' => 'admin',
+        ],
+        '#default_value' => $entity,
+      ],
+      '_weight' => [
+        '#type' => 'weight',
+        '#title_display' => 'invisible',
+        '#delta' => $weight_delta,
+        '#default_value' => $delta,
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+    $parent_values = $values;
+
+    // Get the groups from the other groups widget.
+    foreach ($form[$this->fieldDefinition->getName()]['other_groups'] as $key => $value) {
+      if (!is_int($key)) {
+        continue;
+      }
+
+      // Matches the entity label and ID. E.g. 'Label (123)'. The entity ID will
+      // be captured in it's own group, with the key 'id'.
+      preg_match("|.+\((?<id>[\w.]+)\)|", $value['target_id']['#value'], $matches);
```

Comment:
@amitaibu Comment added. I added a named capture group too, for more readability with how the regex values are used. What do you think?